### PR TITLE
[#536] Add support for custom private key file for public-key export and public-key export-address commands

### LINKED ISSUE #536: Support customer private key file with public-key export subcommand
Besu can be run with a custom private key using `--node-private-key-file` but that argument is not supported by the `public-key export` subcommand. It should be.

Acceptance Criteria
 * `besu public-key export --node-private-key-file <keyfile>` argument exports the public key associated with the private key in `<keyfile>`

### PR DESCRIPTION
**Pull request by diega**
## PR description
This PR adds `--node-private-key-file` option to `public-key export-address` and `public-key export` commands.

## Fixed Issue(s)
Fixes #536

## Changelog

- [ ] Add support for custom private key file for public-key export and public-key export-address commands

### CODE REVIEW

File: besu/src/test/java/org/hyperledger/besu/cli/PublicKeySubCommandTest.java [hunk 1]
```diff
@@ -141,6 +177,30 @@ public void callingPublicKeyExportSubCommandWithFilePathMustWritePublicKeyInThis
     assertThat(commandErrorOutput.toString()).isEmpty();
   }
 
+  @Test
+  public void callingPublicKeyExportSubCommandWithPrivateKeyFileMustWriteKeyToStandardOutput()
```
Comment (atoulme) [inline besu/src/test/java/org/hyperledger/besu/cli/PublicKeySubCommandTest.java:181]: What is the behavior if the file doesn't exist or cannot be read?

Comment (diega) [reply besu/src/test/java/org/hyperledger/besu/cli/PublicKeySubCommandTest.java:183]: It throws an exception. I just added a test that reflects that behavior

**Comment by macfarla** [comment]
org.hyperledger.besu.Besu public-key export-address --node-private-key-file=/nope/issue-536.key
picocli.CommandLine$ExecutionException: Error while running command (org.hyperledger.besu.cli.subcommands.PublicKeySubCommand$AddressSubCommand@5f3b9c57): java.lang.IllegalArgumentException: Cannot store generated private key.
	at picocli.CommandLine.executeUserObject(CommandLine.java:1948)
	at picocli.CommandLine.access$1300(CommandLine.java:145)
	at picocli.CommandLine$RunLast.executeUserObjectOfLastSubcommandWithSameParent(CommandLine.java:2352)
	at picocli.CommandLine$RunLast.handle(CommandLine.java:2346)
	at picocli.CommandLine$RunLast.handle(CommandLine.java:2311)
	at picocli.CommandLine$AbstractParseResultHandler.handleParseResult(CommandLine.java:2172)
	at picocli.CommandLine.parseWithHandlers(CommandLine.java:2550)
	at org.hyperledger.besu.cli.util.ConfigOptionSearchAndRunHandler.handle(ConfigOptionSearchAndRunHandler.java:53)
	at org.hyperledger.besu.cli.util.ConfigOptionSearchAndRunHandler.handle(ConfigOptionSearchAndRunHandler.java:31)
	at picocli.CommandLine$AbstractParseResultHandler.handleParseResult(CommandLine.java:2172)
	at picocli.CommandLine.parseWithHandlers(CommandLine.java:2550)
	at org.hyperledger.besu.cli.BesuCommand.parse(BesuCommand.java:1363)
	at org.hyperledger.besu.cli.BesuCommand.parse(BesuCommand.java:1192)
	at org.hyperledger.besu.Besu.main(Besu.java:50)
Caused by: java.lang.IllegalArgumentException: Cannot store generated private key.
	at org.hyperledger.besu.crypto.KeyPairUtil.storeKeyFile(KeyPairUtil.java:91)
	at org.hyperledger.besu.crypto.KeyPairUtil.loadKeyPair(KeyPairUtil.java:73)
	at org.hyperledger.besu.cli.BesuCommand.loadKeyPair(BesuCommand.java:1329)
	at org.hyperledger.besu.cli.subcommands.PublicKeySubCommand$AddressSubCommand.run(PublicKeySubCommand.java:168)
	at picocli.CommandLine.executeUserObject(CommandLine.java:1939)
	... 13 more

Process finished with exit code 1

File: besu/src/main/java/org/hyperledger/besu/cli/options/stable/NodePrivateKeyFileOption.java [hunk 2]
```diff
@@ -0,0 +1,39 @@
+/*
+ * Copyright ConsenSys AG.
```
Comment (macfarla) [inline besu/src/main/java/org/hyperledger/besu/cli/options/stable/NodePrivateKeyFileOption.java:2]: nit: new files should have * Copyright contributors to Hyperledger Besu  https://wiki.hyperledger.org/display/TSC/Copyright+and+License+Policy

**Review by macfarla — CHANGES_REQUESTED** [review]
When testing this locally, if I specify a non-existent path, I get an exception.
If I specify an existent path but a file that doesn't exist, a `key` file is generated in the supplied path - see comments

Would be nice to add tests for this.

**Comment by diega** [comment]
Hello @macfarla, thanks for the review. I think I addressed your comments.
I've added a check for the existence of the `--node-private-key-file` so I can throw a  `ParameterException` as you suggested. With this I think there is no need to cover the case for the `key` file creation for the case when the file doesn't exists but the parent directory does.
I've also added a test for when the file doesn't contain a valid private key.

File: besu/src/main/java/org/hyperledger/besu/cli/subcommands/PublicKeySubCommand.java [hunk 3]
```diff
@@ -153,34 +124,52 @@ private void outputPublicKey(final NodeKey nodeKey) {
...
+      }
+
+      final KeyPair keyPair;
+      try {
+        keyPair = besuCommand.loadKeyPair(nodePrivateKeyFileOption.getNodePrivateKeyFile());
+      } catch (IllegalArgumentException e) {
+        throw new CommandLine.ParameterException(
+            spec.commandLine(), "Private key cannot be loaded from file", e);
+      }
+      String output = outputFunction.apply(keyPair);
```
Comment (macfarla) [inline besu/src/main/java/org/hyperledger/besu/cli/subcommands/PublicKeySubCommand.java:162]: variable can be final

File: besu/src/main/java/org/hyperledger/besu/cli/subcommands/PublicKeySubCommand.java [hunk 4]
```diff
@@ -153,34 +124,52 @@ private void outputPublicKey(final NodeKey nodeKey) {
...
+        final File exportFile, final Function<KeyPair, String> outputFunction) {
+      checkNotNull(parentCommand);
+      BesuCommand besuCommand = parentCommand.parentCommand;
+      checkNotNull(besuCommand);
 
-      // if we have an output file defined, print to it
-      // otherwise print to standard output.
-      if (addressExportFile != null) {
-        final Path path = addressExportFile.toPath();
+      File nodePrivateKeyFile = nodePrivateKeyFileOption.getNodePrivateKeyFile();
```
Comment (macfarla) [inline besu/src/main/java/org/hyperledger/besu/cli/subcommands/PublicKeySubCommand.java:149]: final

File: besu/src/main/java/org/hyperledger/besu/cli/options/stable/NodePrivateKeyFileOption.java [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+
+  public static NodePrivateKeyFileOption create() {
+    return new NodePrivateKeyFileOption();
+  }
+
+  @CommandLine.Option(
+      names = {"--node-private-key-file"},
+      paramLabel = MANDATORY_PATH_FORMAT_HELP,
+      description =
+          "The node's private key file (default: a file named \"key\" in the Besu data folder)")
```
Comment (macfarla) [inline besu/src/main/java/org/hyperledger/besu/cli/options/stable/NodePrivateKeyFileOption.java:33]: nit: s/folder/directory

**Review by macfarla — APPROVED** [review]
couple of small comments, overall LGTM!

**Comment by macfarla** [comment]
Also @diega your changelog entry is good - can you add it to the CHANGELOG.md file under "Additions"
> Add support for custom private key file for public-key export and public-key export-address commands

**Comment by macfarla** [comment]
All merged in - thanks @diega for your contribution!

**Pull request merged**